### PR TITLE
Fixed disabled users query

### LINKED ISSUE #8008: User count for groups with disabled users is wrong
* having a group "A" with two users "user1" and "user2"
* disable "user1"
* expected: group count for group "A" is 2 and shows the disabled user or it is 1 and doesn't show the user
* actual: group count is 2 and doesn't show the user

bildschirmfoto 2018-01-23 um 15 17 38
bildschirmfoto 2018-01-23 um 15 17 43

cc @michag86 as you also experience this

cc @blizzz @nickvergessen @rullzer

### LINKED ISSUE #8009: Disabled users count for sub admins wrong
* having a group "A" with two users "user1" and "user2"
* having users "user3" and "user4" without a group
* disable "user1" and "user3"
* make "user2" the group admin of group "A"
* login as "user2"
* expected: disabled count is 1 (because only one of the users that are in the scope of the group admin "user2" is disabled)
* actual: disabled count is 2 (user1 and user3)

admin view:
bildschirmfoto 2018-01-23 um 15 46 53

bildschirmfoto 2018-01-23 um 15 46 55

subadmin view:
bildschirmfoto 2018-01-23 um 15 46 57

cc @michag86 for noticing this

@nickvergessen @rullzer @danxuliu @blizzz for the user management 

May be related to #8008

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix #8009 
Fix #8008 
I feel like it's safer to return a different entry instead of directly subtracting the disabled users to the total user count. People might want to get the full data anyway.

### CODE REVIEW

**Comment by skjnldsv** [comment]
Tests ok, ready to review :)

File: lib/private/User/Manager.php [hunk 1]
```diff
@@ -441,6 +441,31 @@ public function countDisabledUsers() {
...
 	}
 
+	/**
+	 * returns how many users are disabled in the requested groups
+	 *
+	 * @param array $groups groupids to search
+	 * @return int
+	 * @since 14.0.0
+	 */
+	public function countDisabledUsersOfGroups(array $groups) {
```
Comment (MorrisJobke) [inline lib/private/User/Manager.php]: Also give a hint for the return type here? `: int` ;)

File: lib/public/IUserManager.php [hunk 2]
```diff
@@ -162,13 +162,22 @@ public function countUsers();
...
 	public function countDisabledUsers();
 
+	/**
+	 * returns how many users are disabled in the requested groups
+	 *
+	 * @param array $groups groupids to search
+	 * @return int
+	 * @since 14.0.0
+	 */
+	public function countDisabledUsersOfGroups(array $groups);
```
Comment (MorrisJobke) [inline lib/public/IUserManager.php]: Same here - type hint for the return type would be nice.

File: lib/private/User/Manager.php [hunk 3]
```diff
@@ -441,6 +441,31 @@ public function countDisabledUsers() {
...
+	 *
+	 * @param array $groups groupids to search
+	 * @return int
+	 * @since 14.0.0
+	 */
+	public function countDisabledUsersOfGroups(array $groups):int {
+		$queryBuilder = \OC::$server->getDatabaseConnection()->getQueryBuilder();
+		$queryBuilder->select($queryBuilder->createFunction('COUNT(Distinct uid)'))
+			->from('preferences', 'p')
+			->innerJoin('p', 'group_user', 'g', 'p.userid = g.uid')
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: This hard codes the user manager to our group backend :x:

File: lib/private/User/Manager.php [hunk 4]
```diff
@@ -441,6 +441,31 @@ public function countDisabledUsers() {
...
+			->where($queryBuilder->expr()->eq('appid', $queryBuilder->createNamedParameter('core')))
+			->andWhere($queryBuilder->expr()->eq('configkey', $queryBuilder->createNamedParameter('enabled')))
+			->andWhere($queryBuilder->expr()->eq('configvalue', $queryBuilder->createNamedParameter('false'), IQueryBuilder::PARAM_STR))
+			->andWhere($queryBuilder->expr()->in('gid', $queryBuilder->createNamedParameter(implode(',', $groups))));
```
Comment (nickvergessen) [inline lib/private/User/Manager.php]: no need to implode, `$queryBuilder->createNamedParameter($groups, IQueryBuilder::PARAM_STR_ARRAY)` has you covered

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: lib/public/IUserManager.php [hunk 5]
```diff
@@ -162,12 +162,21 @@ public function countUsers();
...
+
+	/**
+	 * returns how many users are disabled in the requested groups
+	 *
+	 * @param array $groups groupids to search
+	 * @return int
+	 * @since 14.0.0
 	 */
-	public function countDisabledUsers();
+	public function countDisabledUsersOfGroups(array $groups):int;
```
Comment (MorrisJobke) [inline lib/public/IUserManager.php]: Could you add a space before the `int`? That's how we use it in all other places in our code.

File: lib/public/IUserManager.php [hunk 6]
```diff
@@ -162,12 +162,21 @@ public function countUsers();
 	public function callForAllUsers(\Closure $callback, $search = '');
 
 	/**
-	 * returns how many users have logged in once
+	 * returns how many users are disabled
 	 *
 	 * @return int
-	 * @since 11.0.0
+	 * @since 12.0.0
```
Comment (MorrisJobke) [inline lib/public/IUserManager.php]: Why isthmus updated?

Comment (skjnldsv) [reply lib/public/IUserManager.php]: it was an error, implemented in 12 in the private fct but in 11 in the public interface :)
I guess a copy/paste error

File: lib/private/Group/Backend.php [hunk 7]
```diff
@@ -39,6 +39,7 @@
 		self::ADD_TO_GROUP => 'addToGroup',
 		self::REMOVE_FROM_GOUP => 'removeFromGroup',
 		self::COUNT_USERS => 'countUsersInGroup',
+		self::COUNT_DISABLED => 'countDisabledInGroup',
```
Comment (MorrisJobke) [inline lib/private/Group/Backend.php]: Not defined 🙈

Comment (skjnldsv) [reply lib/private/Group/Backend.php]: Yeah, we discussed it with @rullzer and @nickvergessen.
Unfortunately https://github.com/nextcloud/server/blob/da6c2c9da1721de7aa05b15af1356e3511069980/lib/public/GroupInterface.php#L48-L59 is full :(

Comment (MorrisJobke) [reply lib/private/Group/Backend.php]: We should go the interface way 😉 see #8935

Comment (nickvergessen) [reply lib/private/Group/Backend.php]: that was also the conclusion in IRC ;)

**Comment by skjnldsv** [comment]
Let's close this in favor of a new one

**Pull request closed without merging**